hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
  * License along with this library; if not, see <http://www.gnu.org/licenses/>.
  */
 
-#pragma once
+#ifndef BLE_HAL_H
+#define BLE_HAL_H
 
-#include "ble_hal_impl.h"
+#include "hal_platform.h"
 
-#include "system_error.h"
+#if HAL_PLATFORM_BLE
 
 #include <stdint.h>
 #include <stddef.h>
 
-// TODO: This API is not ready for general purpose usage
+#include "ble_hal_defines.h"
+#include "ble_hal_impl.h"
+
+/**
+ * @addtogroup ble_api
+ *
+ * @brief
+ *   This module provides BLE HAL APIs.
+ * @{
+ */
+
+typedef enum {
+    BLE_ANT_DEFAULT  = 0,
+    BLE_ANT_INTERNAL = 1,
+    BLE_ANT_EXTERNAL = 2,
+} hal_ble_ant_type_t;
+
+/* BLE device address */
+typedef struct {
+    uint8_t addr_type;
+    uint8_t addr[BLE_SIG_ADDR_LEN];
+} hal_ble_addr_t;
+
+/* BLE UUID, little endian */
+typedef struct {
+    uint8_t type;
+    union {
+        uint16_t uuid16;
+        uint8_t uuid128[BLE_SIG_UUID_128BIT_LEN];
+    };
+} hal_ble_uuid_t;
+
+/* BLE advertising parameters */
+typedef struct {
+    uint8_t version;
+    uint8_t type;
+    uint8_t filter_policy;
+    uint16_t interval;                  /**< Advertising interval in 625 us units. */
+    uint16_t timeout;                   /**< Advertising timeout in 10 ms units.*/
+    uint8_t inc_tx_power;
+} hal_ble_adv_params_t;
+
+/* BLE scanning parameters */
+typedef struct {
+    uint8_t version;
+    uint8_t active;
+    uint8_t filter_policy;
+    uint16_t interval;                  /**< Scan interval in 625 us units. */
+    uint16_t window;                    /**< Scan window in 625 us units. */
+    uint16_t timeout;                   /**< Scan timeout in 10 ms units. */
+} hal_ble_scan_params_t;
+
+/* BLE connection parameters */
+typedef struct {
+    uint8_t  version;
+    uint16_t min_conn_interval;         /**< Minimum Connection Interval in 1.25 ms units.*/
+    uint16_t max_conn_interval;         /**< Maximum Connection Interval in 1.25 ms units.*/
+    uint16_t slave_latency;             /**< Slave Latency in number of connection events.*/
+    uint16_t conn_sup_timeout;          /**< Connection Supervision Timeout in 10 ms units.*/
+} hal_ble_conn_params_t;
+
+/* BLE service definition */
+typedef struct {
+    uint8_t version;
+    uint16_t start_handle;
+    uint16_t end_handle;
+    hal_ble_uuid_t uuid;
+} hal_ble_svc_t;
+
+typedef struct {
+    uint8_t version;
+    uint8_t properties;
+    uint16_t service_handle;
```

Comment:
For better portability to other platforms in the future consider using a separate typedefed type (possible overridable by the implementation) for handles everywhere. `uint16_t` may look like a good choice for nRF52840, but for other platforms opaque `void*` may be preferred for example.

This will also improve type safety and readability.